### PR TITLE
[DRAFT] Add Docker integeration and documentation

### LINKED ISSUE #8: Help Dockerizing Lute
Hello all, if you're reading this, perhaps you have Docker experience and can help Dockerize Lute.

I've had a few people ask about having Lute Dockerized. It's been a loooong time since I've hacked on Docker, and I don't have the time or energy to spend fiddling with it. Perhaps someone is looking for a Docker project to try out, and would be willing to contribute.

Below is some info, LMK if you need more. This is a free project, I'm not looking to make money from it, and so I can't pay anyone. Hopefully this doesn't make you feel you're being taken advantage of ... (as a dev, I used to always feel that certain types of ppl were leeching off me) ... my limited dev time is better spent working on Lute itself.

Cheers and best wishes!

Jeff

## Specs for Lute:

* needs Apache (due to using Apache rewrites for URLs)
* PHP 8.1 (and maybe later)
* MySQL 8.0.31

## More notes:

* Lute users install and run Lute on their own machines. I'm not running a business, I just provide this free tool for language learners to use. So, this docker installation is for people to run on their own machines ... and perhaps it would be nice to have a Docker development environment as well. Your input appreciated.  Users would be expected to have the necessary bits for Docker on their machines (Docker desktop or similar, docker compose, etc)
* The project uses Symfony, a PHP framework, and php composer to pull in all of the packages.  I am not sure if the code and framework should be built directly into the container, or if they should be mounted from an external directory ... your input appreciated.
* The database migrations are managed automagically by the project -- Lute is given root access to the MySQL so that it can create tables, etc.
* The project needs a single configuration file to describe the database connection params. Maybe these can be passed in as env params to the container -- or perhaps just the file is somehow passed in?
* The project also stores some downloaded images in a /public/usermedia directory, so those would need to be ... mounted? Along with the db?
* Users should be able to update their Lute by running the latest container (at which point the db migrations are automatically applied), or perhaps replacing the mounted code directory with newer code (and then somehow killing the cache in the container ... might get annoying)
* The project build is single-click. CI is running locally and on GitHub.

### PR DESCRIPTION
**Pull request by AdamHebby**
WIP - Need to figure out how to run tests via Docker & update documentation. Seems like some mangling of env files may be needed.

Fixes #8

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -1,4 +1,4 @@
-DB_HOSTNAME=localhost
+DB_HOSTNAME=lutedb
```
Comment (AdamHebby) [inline .env:1]: TODO: Check if this is really needed. Ideally don't want to make yet another env file, there's enough of those already ;)

Comment (jzohrab) [reply .env:1]: Maybe we could create a separate .env file just for the Docker build, stored in /docker.

File: docker/app-init.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# todo: replace prod with $1 and pass in from docker-compose.yml from APP_ENV
+
+cd /var/www/html && \
+    composer run-script db:setup:prod && \
+    composer run-script db:migrate:prod
```
Comment (jzohrab) [inline docker/app-init.sh:7]: `db:setup:prod` calls the script `src/Utils/MigrationHelper.php -> doSetup()`, which should run all of the pending migrations already ... the `db:migrate:prod` might not be necessary.

Comment (jzohrab) [reply docker/app-init.sh:7]: ps - the notes I'm adding here are as much for my benefit as for yours. :-)

File: docs/installation/install_docker.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# Installing via Docker
+
+## Get Docker
+
+Docker is a containerization platform that allows you to run applications in a sandboxed environment. You can install Docker on your local machine or on a server. For more information, see the [Docker documentation](https://docs.docker.com/).
+
+You can download docker from here: [Get Docker](https://docs.docker.com/get-docker/).
```
Comment (jzohrab) [inline docs/installation/install_docker.md:7]: For mac, Docker desktop was brutal.  I switched to colima, it seems to be better.

Hard to say what docs are necessary for Docker on the whole.

These docs will go in the wiki :-)  Thanks for this doc!

**Comment by jzohrab** [comment]
Running this on my mac gives me an error:

```<li>No such file or directory</li>```

Interestingly, the log also says `symfony  | Setting up test_lute` even though  I changed the .env file to `DB_DATABASE=zzlutedocker` -- so the setup is still taking the `.env.local` file, which does have `DB_DATABASE=test_lute`.

<details>
  <summary>full trace</summary>

```
MacBook-Air-2:lute_dev jeff$ docker-compose up
[+] Running 3/3
 ⠿ Network lute_dev_default  Created                                                           0.6s
 ⠿ Container lutedb          Created                                                           1.8s
 ⠿ Container symfony         Created                                                           1.0s
Attaching to lutedb, symfony
lutedb   | 2023-03-08 03:25:03+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.31-1.el8 started.
lutedb   | 2023-03-08 03:25:08+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
lutedb   | 2023-03-08 03:25:08+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.31-1.el8 started.
lutedb   | '/var/lib/mysql/mysql.sock' -> '/var/run/mysqld/mysqld.sock'
lutedb   | 2023-03-08T03:25:49.193324Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
lutedb   | 2023-03-08T03:25:49.387667Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.31) starting as process 1
lutedb   | 2023-03-08T03:25:49.668681Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
lutedb   | 2023-03-08T03:26:31.903694Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
lutedb   | 2023-03-08T03:26:53.533435Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
lutedb   | 2023-03-08T03:26:53.537146Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
lutedb   | 2023-03-08T03:26:53.564581Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users. Consider choosing a different directory.
lutedb   | 2023-03-08T03:26:54.058625Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
lutedb   | 2023-03-08T03:26:54.059610Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.31'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
symfony  | > APP_ENV=prod php db/setup.php
symfony  | 
symfony  | Setting up test_lute
symfony  | 
symfony  | Done.
symfony  | ERROR: 
symfony  | <!DOCTYPE html>
symfony  | <html>
symfony  |   <head>
symfony  |     <meta http-equiv="content-type" content="text/html; charset=utf-8" />
symfony  |     <title>LUTE - fatal error</title>
symfony  |     <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"/>
symfony  | 
symfony  |     <link rel="stylesheet" type="text/css" href="/css/styles.css" />
symfony  |     </script>
symfony  |   </head>
symfony  | 
symfony  |   <body>
symfony  | 
symfony  |     <h1>Hmmm, something has gone wrong ...</h1>
symfony  | 
symfony  |     <ul>
symfony  |             <li>No such file or directory</li>
symfony  |           </ul>
symfony  | 
symfony  |     <p>If the above look like database errors, please do the following:</p>
symfony  |     <ul>
symfony  |       <li>
symfony  |         Make sure that MySQL is running!
symfony  |       </li>
symfony  |       <li>
symfony  |         Check the information in your <b>.env.local</b>
symfony  |       </li>
symfony  |       <li>
symfony  |         Ensure that you're using a root db account, with proper privileges.
symfony  |       </li>
symfony  |     </ul>
symfony  |   </body>
symfony  | 
symfony  | </html>
symfony  | 
```

The above then keeps repeating until I ctl-C.

</details>

**Comment by jzohrab** [comment]
Hi @AdamHebby - I was looking at #11 instead of this one.  It looks like #11 was based off of this one.

**Comment by jzohrab** [comment]
Closing in favor of #11, which I believe incorporates this work somewhat.

**Pull request closed without merging**